### ISSUE TITLE
Session expiration issues / silent authentication support?

### ISSUE THREAD

**Issue by fotinakis** [auth0-community/ember-simple-auth-auth0#120]
Hello friends,

It looks like [this recent bugfix](https://github.com/auth0-community/ember-simple-auth-auth0/commit/5270123aad39bfa3beb29e67c846396b34fb1765) uses the JWT token expiration time to decide whether to restore the session, whereas before the session was restored no matter what.

Though it feels like this bugfix is likely a good change, it has had the very unfortunate side-effect in our Ember SPA of automatically logging users out every 1 hour (due to JWT Expiration settings which we had slightly tuned down, the default auth0 setting for this is 36000 seconds, or 10 hours). After this expiration, users are forced through the modal flow where they have to select the account they used last - ie. the modal with "Last time you logged in with" and you click the account to re-login.

This has created a very large UI/UX annoyance because it feels to users like they are getting logged out super frequently and have re-login every day. We've had numerous customers report this to us directly in the last week after we upgraded ember-simple-auth-auth0, and we only recently identified the JWT enforcement change as the culprit.

The JWT Expiration time should be very short, for security concerns, so increasing this to some arbitrary large value is not an option.

Based on Auth0's [Token Lifetime docs](https://auth0.com/docs/tokens/id-token#token-lifetime) I believe the fix for this in SPA's with auth0 is a "silent authentication" technique:

> This method is not an option for Single Page Apps (SPAs), since for security reasons you cannot get a refresh_token from the Implicit Grant (the OAuth flow typically used from Client-side Web Apps). In that case you would have to use [silent authentication](https://auth0.com/docs/api-auth/tutorials/silent-authentication).

--

So, my questions are:

- Are there plans to support silent authentication in ember-simple-auth-auth0? This would automatically refresh the id_token underneath and users wouldn't notice.
- Is there a workaround in the meantime? Maybe providing a way to use the old resolve-always behavior?

**Comment by XaserAcheron** [comment]
Having native silent authentication support in the addon is definitely the way to go. There's two halves, basically:
- Trigger Silent Auth during session restore.
- Trigger Silent Auth when the current session expires, so users aren't forcibly logged out.

Related to the above, I'm planning on adding a `sessionExpired` hook to applicationRouteMixin that allows for custom behavior once the token goes bad, since the addon's current behavior (always force a log-out) is pretty lacking, frankly.

For an immediate workaround, you can restore the old behavior by creating a new authenticator that inherits from the `auth0-lock` authenticator (or `auth0-lock-passwordless` if you're using that), then overriding the `restore` method like so:

```js
// app/authenticators/auth0-lock.js

import Ember from 'ember';
import Auth0LockAuthenticator from 'ember-simple-auth-auth0/authenticators/auth0-lock';

const { RSVP } = Ember;

export default Auth0LockAuthenticator.extend({
  restore(data) {
    return RSVP.resolve(data);
  }
});
```

I'm not entirely sure this will work as expected, though, since there's a timeout function that occurs shortly after that detects token expiration and forces a log-out anyway. I'm not sure how you were handling this case before, unless you were using the base `ember-simple-auth` applicationRouteMixin (a few folks were using this to work around some old bugs in this addon's mixin, turns out).

**Comment by XaserAcheron** [comment]
Both -- at least that's the plan. I've got a working Silent Auth workflow in my application that uses this addon, so it's _hopefully_ just a matter of backporting it. We'll see, I suppose. :P

**Comment by queenvictoria** [comment]
I'd love to get started on this one and submit a PR if anyone has any guidance as to how to begin.

**Comment by XaserAcheron** [comment]
Sorry for the long wait -- finally started on this.

Still a few things to do, but here's where I'm at so far:
- Added an `auth0-silent-auth` authenticator. It needs to be invoked manually at present, but it works.
- Turns out there was already a `beforeSessionExpired` hook in `applicationRouteMixin`, so there's no action required there. If you override this and return `RSVP.reject()`, this will stop the live session from expiring (use at your own risk though).

Next steps:
- Automatically perform silent authentication when token expires during use.
- Automatically perform silent authentication when session data restore fails due to token expiration.
- Add some `environment.config` options for the automatic silentAuth calls, since it's necessary to pass some things in (scope, responseType, et.al) to get it to work.
- Add a `sessionRenewing` hook to `applicationRouteMixin`, since there's a brief window of time in which the current session is invalid and we're waiting for the new session token to arrive, and apps will want to handle this.
- Document stuff.

The automatic silentAuth stuff will be opt-in via `environment.config`, but the goal is to otherwise have it Just Work(tm) without any fuss. Fingers crossed!

**Comment by XaserAcheron** [comment]
I haven't forgotten about this, but I went ahead and accelerated v4.1.0's release since #129 was tripping folks up. I've changed the milestone accordingly, though the intent isn't to postpone it indefinitely or anything.

**Comment by queenvictoria** [comment]
Thank you! I was going to try and push ahead your branch today. I'm starting from cold. Any tips?

**Comment by XaserAcheron** [comment]
I'm kinda rubbish at explanations so I'm not entirely sure what to suggest offhand. I left a sort of to-do list above, so perhaps start with one of those?

An unusual thing, I suppose, is that the repo is set up to use GitFlow with the [HubFlow](https://datasift.github.io/gitflow/) extension. It's not strictly necessary to get it set up to make a PR or anything, just remember to work from the `develop` branch, not `master`.

Thanks either way! Apologies for being a bit of an odd guide.

**Comment by queenvictoria** [comment]
Ok cool. I was going to start from your feature branch (which sounds gitflow-y). But I'll read up on HubFlow first.

**Comment by queenvictoria** [comment]
Hey @XaserAcheron I'm having difficulty getting the tests and the dummy working. Maybe next week we could find a quiet channel and try to resolve? As I'd like to spend time addressing the issue not the build ;)

**Comment by fotinakis** [comment]
Any update here?

**Comment by fotinakis** [comment]
@XaserAcheron Friendly ping for a conversation here. This is becoming a major issue for us and causing lots of customer reports about the frustration of frequently being logged out.

We did add back the old behavior by creating a custom authenticator that restores the session, but I think we are hitting the issue you mentioned with the timeout function that occurs shortly after that detects token expiration and forces a log-out anyway.

Depending on your expected timeline we might be able to dedicate some time to implement this with some direction, but I fear we might hit the same snags @queenvictoria did above in getting set up.

Anything we can do to help sponsor this moving forward?

**Comment by fotinakis** [comment]
We've moved forward and implemented silent auth directly in our app. Here it is for reference: https://github.com/percy/percy-web/pull/691

**Comment by XaserAcheron** [comment]
@fotinakis Sorry, I missed your ping -- was out of the house last night.

The missing pieces on this end are adding all the automatic hooks for triggering silent authentication. I _think_ we can avoid that ten-second delay & background update by hooking everything up to the usual promise workflow -- basically do a sort of "try to do silentAuth; if it fails, do something else" kind of thing -- but it seems like this'll take a bit of care to make sure failures are handled OK.

Seems like the fire's out, anyway (once that PR's accepted, that is). Anything I can help with in the meantime aside from finishing this up?

**Comment by XaserAcheron** [comment]
Just finished up a first draft and published `v4.2.0-beta.0`. @fotinakis, interested in trying it out?

There's some docs on how to use it in #137 -- I'll put these in the readme once everything's all kosher. The important gist is with all 3 settings on, silent authentication is only attempted in one of 3 cases:
- On app load, but _only_ if trying to restore an otherwise-valid session with an expired token.
- On a regular, app-specified timer (e.g. every 30 minutes in the linked test configuration).
- Just-in-time when the token expires.

It's all opt-in, and it ought not be spamming silent auth calls in any case. Though a beta's a beta, so let me know if anything seems goofy!

I'll keep this issue open for a bit longer just to confirm everything's A-OK.

**Comment by fotinakis** [comment]
Hey @XaserAcheron — nice, excited to try this out. Out of curiosity, is there a timing problem with "just-in time" token expiration? If the id_token has expired, the Auth0 session will still allow generating new id_tokens? (I assume the Auth0 SSO session is separate and that it'll be fine, just making sure).

**Comment by XaserAcheron** [comment]
Shoot! I was going to add support for `redirectUrl` functions, but I totally forgot. Thanks for the reminder!

Since it's defined in `environment.js`, which doesn't have any visibility to the live app, we'll need to pass something in as a parameter for the function to be of any use. Any ideas on what the param should be? You've got better insight into this than me, I think.

I'm not totally sure about the just-in-time question; there's presently a gap while waiting for the silentAuth call to finish where the active session token is expired (not quite sure how best to handle it just yet), but in the worst case, I think all you'd get is a "sorry, I can't do that" message when trying to consume the expired token. Still an error, but a fail-closed sort of thing.

Anyhow, good to hear it's mostly working! Not far from the finish line, I suppose.

**Comment by fotinakis** [comment]
Ah yeah, I just tried it out and I'm not sure if it's going to be possible to have a function declared in environment.js make it into the runtime app at all — they are evaluated in node world at build time by ember-cli, and functions are stripped from the compiled version.

The use case for us (and probably a common one) is that we just need to append a path to `window.location.origin` like:

```
redirectUri: `${window.location.origin}/api/auth/auth0/callback`,
```

So one option is you could provide a `redirectUriPath` option which is set like:

```
redirectUriPath: '/api/auth/auth0/callback',
```
and just document that `window.location.origin` will be prepended to the path arg, and that only one of `redirectUriPath` or `redirectUri` should be given. That would solve our particular case.

**Comment by fotinakis** [comment]
@XaserAcheron After some more testing I've found two more observations (one critical, one general):

1) *critical* if a user has third-party cookies disabled, silentAuth breaks the normal login flow — `ember_simple_auth-session` in localstorage becomes `{"authenticated":{}}` and the user is logged out. (Per the Auth0 docs, if a user has third-party cookies disabled or running in Safari ITP mode, Silent Auth will never work).

2) *non-critical* on first login, silent auth background jobs don't seem to be running (no network requests are made). After session restore on refresh, the background jobs start.

**Comment by fotinakis** [comment]
Can confirm that the remaining blockers here are:
1) Support a `redirectUriPath` option, which becomes `redirectUri` with the current window.location.origin prepended.
2) Allow Silent Auth to fail without breaking normal auth (ie. disable third-party cookies, and normal login should still work).

**Comment by XaserAcheron** [comment]
Thanks for the feedback -- not sure what's up with the empty session stuff. Definitely a bug of some sort in the addon.

As a heads-up, I probably won't have a chance to work on this until Friday at the earliest -- just got a mountain of work dropped in my lap today that'll eat up my immediate time. :(

**Comment by fotinakis** [comment]
Hey @XaserAcheron — I was actually unable to recreate the "breaking normal auth" case, I've now tested it under conditions where third-party cookies are disabled and things work normally for us.

I've implemented `redirectPath` here and tested this extensively in our environment:
https://github.com/auth0-community/ember-simple-auth-auth0/pull/138

**Comment by XaserAcheron** [comment]
Huh, strange -- I'll keep an eagle eye out for that phantom breakage. Would be unfortunate to release a ghost. ;)

Seems like most of the loose ends are all tied up, so I'll go ahead and mark this one done/closed/whatnot. If any new troublesome spots arise, we'll open up a new issue and continue there.

Thanks again for all the help!

**Comment by fotinakis** [comment]
Thanks @XaserAcheron. Can you release a new beta version, or when is 4.2.0 coming out?

**Comment by XaserAcheron** [comment]
Oops! I thought I forgot something. :P

Just released `v4.2.0-beta.1` -- if nothing else crops up in the near future, it's all ready for release, I believe.
